File: app/src/main/java/com/qltech/bws/BillingOrderModule/Activities/PaymentActivity.java
package com.qltech.bws.BillingOrderModule.Activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.qltech.bws.AddPayment.AddPaymentActivity;
import com.qltech.bws.BWSApplication;
import com.qltech.bws.BillingOrderModule.Models.CardListModel;
import com.qltech.bws.BillingOrderModule.Models.CardModel;
import com.qltech.bws.BillingOrderModule.Models.PayNowDetailsModel;
import com.qltech.bws.BillingOrderModule.Models.PlanListBillingModel;
import com.qltech.bws.MembershipModule.Activities.OrderSummaryActivity;
import com.qltech.bws.MembershipModule.Models.MembershipPlanListModel;
import com.qltech.bws.R;
import com.qltech.bws.Utility.APIClient;
import com.qltech.bws.Utility.CONSTANTS;
import com.qltech.bws.databinding.ActivityPaymentBinding;
import com.qltech.bws.databinding.CardsListLayoutBinding;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.qltech.bws.BillingOrderModule.Activities.MembershipChangeActivity.renewPlanFlag;
import static com.qltech.bws.BillingOrderModule.Activities.MembershipChangeActivity.renewPlanId;
import static com.qltech.bws.BillingOrderModule.Fragments.CurrentPlanFragment.PlanStatus;
import static com.qltech.bws.BillingOrderModule.Fragments.CurrentPlanFragment.invoicePayId;

public class PaymentActivity extends AppCompatActivity {
    ActivityPaymentBinding binding;
    AllCardsAdapter adapter;
    Context context;
    String card_id, userId, TrialPeriod, comeFrom = "";
    int position;
    ArrayList<PlanListBillingModel.ResponseData.Plan> listModelList2;
    private ArrayList<MembershipPlanListModel.Plan> listModelList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_payment);
        context = PaymentActivity.this;

        SharedPreferences shared = context.getSharedPreferences(CONSTANTS.PREF_KEY_LOGIN, Context.MODE_PRIVATE);
        userId = (shared.getString(CONSTANTS.PREF_KEY_UserID, ""));
        Glide.with(context).load(R.drawable.loading).asGif().into(binding.ImgV);
        RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(context);
        binding.rvCardList.setLayoutManager(mLayoutManager);
        binding.rvCardList.setItemAnimator(new DefaultItemAnimator());

        if (getIntent() != null) {
            TrialPeriod = getIntent().getStringExtra("TrialPeriod");
            position = getIntent().getIntExtra("position", 0);
            if (getIntent().hasExtra("comeFrom")) {
                comeFrom = getIntent().getStringExtra("comeFrom");
                listModelList2 = getIntent().getParcelableArrayListExtra("PlanData");
            } else {
                listModelList = getIntent().getParcelableArrayListExtra("PlanData");
            }
        }
        binding.llBack.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent i = new Intent(context, OrderSummaryActivity.class);
                i.putExtra("comeFrom", "membership");
                i.putParcelableArrayListExtra("PlanData", listModelList2);
                i.putExtra("TrialPeriod", "");
                i.putExtra("position", position);
                startActivity(i);
                finish();
            }
        });


        binding.llAddNewCard.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent i = new Intent(context, AddPaymentActivity.class);
                i.putExtra("ComePayment", "2");
                startActivity(i);
            }
        });

        prepareCardList();
    }

    @Override
    public void onResume() {
        super.onResume();
        prepareCardList();
    }

    private void prepareCardList() {
        try {
            if (BWSApplication.isNetworkConnected(context)) {
                showProgressBar();
                Call<CardListModel> listCall = APIClient.getClient().getCardLists(userId);
                listCall.enqueue(new Callback<CardListModel>() {
                    @Override
                    public void onResponse(Call<CardListModel> call, Response<CardListModel> response) {
                        if (response.isSuccessful()) {
                            hideProgressBar();
                            try {
                                CardListModel cardListModel = response.body();
                                if (cardListModel.getResponseCode().equalsIgnoreCase(getString(R.string.ResponseCodesuccess))) {
                                    if (cardListModel.getResponseData().size() == 0) {
                                        binding.rvCardList.setAdapter(null);
                                        binding.rvCardList.setVisibility(View.GONE);
                                    } else {
                                        binding.rvCardList.setVisibility(View.VISIBLE);
                                        adapter = new AllCardsAdapter(cardListModel.getResponseData(), binding.ImgV,
                                                binding.progressBarHolder, binding.rvCardList);
                                        binding.rvCardList.setAdapter(adapter);
                                    }

                                    binding.btnCheckout.setOnClickListener(view -> {
                                        if (cardListModel.getResponseData().size() == 0) {
                                            BWSApplication.showToast("Please enter card details", context);
                                        } else {

                                            if (BWSApplication.isNetworkConnected(context)) {
                                                showProgressBar();
                                                Call<PayNowDetailsModel> listCall = APIClient.getClient().getPayNowDetails(userId, card_id, renewPlanId, renewPlanFlag,
                                                        invoicePayId, PlanStatus);
                                                listCall.enqueue(new Callback<PayNowDetailsModel>() {
                                                    @Override
                                                    public void onResponse(Call<PayNowDetailsModel> call, Response<PayNowDetailsModel> response) {
                                                        if (response.isSuccessful()) {
                                                            hideProgressBar();
                                                            PayNowDetailsModel listModel1 = response.body();
                                                            BWSApplication.showToast(listModel1.getResponseMessage(), context);
                                                            Intent i = new Intent(context, BillingOrderActivity.class);
                                                            startActivity(i);
                                                            finish();
                                                        }
                                                    }

                                                    @Override
                                                    public void onFailure(Call<PayNowDetailsModel> call, Throwable t) {
                                                        hideProgressBar();
                                                    }
                                                });
                                            } else {
                                                hideProgressBar();
                                                BWSApplication.showToast(getString(R.string.no_server_found), context);
                                            }
                                        }
                                    });

                                } else {
                                    BWSApplication.showToast(cardListModel.getResponseMessage(), context);
                                }
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                    }

                    @Override
                    public void onFailure(Call<CardListModel> call, Throwable t) {
                        hideProgressBar();
                    }

                });
            } else {
                BWSApplication.showToast(getString(R.string.no_server_found), context);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public class AllCardsAdapter extends RecyclerView.Adapter<AllCardsAdapter.MyViewHolder> {
        private List<CardListModel.ResponseData> listModelList;
        ImageView ImgV;
        FrameLayout progressBarHolder;
        RecyclerView rvCardList;
        AllCardsAdapter adapter;

        public AllCardsAdapter(List<CardListModel.ResponseData> listModelList, ImageView ImgV,
                               FrameLayout progressBarHolder, RecyclerView rvCardList) {
            this.listModelList = listModelList;
            this.ImgV = ImgV;
            this.progressBarHolder = progressBarHolder;
            this.rvCardList = rvCardList;
        }

        @Override
        public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            Glide.with(context).load(R.drawable.loading).asGif().into(ImgV);
            CardsListLayoutBinding v = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext())
                    , R.layout.cards_list_layout, parent, false);
            return new MyViewHolder(v);
        }

        @Override
        public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
            CardListModel.ResponseData listModel = listModelList.get(position);
            holder.binding.tvCardNo.setText(getString(R.string.first_card_chars) + listModel.getLast4());
            holder.binding.tvExpiryTime.setText("Valid: " + listModel.getExpMonth() + "/" +
                    listModel.getExpYear());
            Glide.with(context).load(listModel.getImage())
                    .thumbnail(0.05f)
                    .crossFade()
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(holder.binding.ivCardimg);
            if (listModel.getIsDefault().equalsIgnoreCase(CONSTANTS.FLAG_ONE)) {
                holder.binding.ivCheck.setImageResource(R.drawable.ic_checked_icon);
                card_id = listModel.getCustomer();
                SharedPreferences shared = getSharedPreferences(CONSTANTS.PREF_KEY_CardID, MODE_PRIVATE);
                SharedPreferences.Editor editor = shared.edit();
                editor.putString(CONSTANTS.PREF_KEY_CardID, card_id);
                editor.commit();
            } else {
                holder.binding.ivCheck.setImageResource(R.drawable.ic_unchecked_icon);
            }

            holder.binding.llAddNewCard.setOnClickListener(view -> {
                if (BWSApplication.isNetworkConnected(context)) {
                    showProgressBar();
                    Call<CardListModel> listCall = APIClient.getClient().getChangeCard(userId, listModel.getCustomer());
                    listCall.enqueue(new Callback<CardListModel>() {
                        @Override
                        public void onResponse(Call<CardListModel> call, Response<CardListModel> response) {
                            hideProgressBar();
                            {
                                CardListModel cardListModel = response.body();
                                if (cardListModel.getResponseCode().equalsIgnoreCase(getString(R.string.ResponseCodesuccess))) {
                                    if (cardListModel.getResponseData().size() == 0) {
                                        rvCardList.setAdapter(null);
                                        rvCardList.setVisibility(View.GONE);
                                    } else {
                                        rvCardList.setVisibility(View.VISIBLE);
                                        adapter = new AllCardsAdapter(cardListModel.getResponseData(), ImgV, progressBarHolder, rvCardList);
                                        rvCardList.setAdapter(adapter);
                                    }
                                } else {
                                    BWSApplication.showToast(cardListModel.getResponseMessage(), context);
                                }

                            }
                        }

                        @Override
                        public void onFailure(Call<CardListModel> call, Throwable t) {
                            hideProgressBar();
                        }

                    });
                } else {
                    BWSApplication.showToast(getString(R.string.no_server_found), context);
                    hideProgressBar();
                }

            });

            holder.binding.rlRemoveCard.setOnClickListener(view -> {
                final Dialog dialog = new Dialog(context);
                dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
                dialog.setContentView(R.layout.cancel_membership);
                dialog.getWindow().setBackgroundDrawable(new ColorDrawable(getResources().getColor(R.color.dark_blue_gray)));
                dialog.getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);

                final TextView tvGoBack = dialog.findViewById(R.id.tvGoBack);
                final TextView tvTitle = dialog.findViewById(R.id.tvTitle);
                final TextView tvSubTitle = dialog.findViewById(R.id.tvSubTitle);
                final RelativeLayout tvconfirm = dialog.findViewById(R.id.tvconfirm);
                tvTitle.setText("Delete payment card");
                tvSubTitle.setText("Are you sure you want to delete the payment card ?");
                dialog.setOnKeyListener((v, keyCode, event) -> {
                    if (keyCode == KeyEvent.KEYCODE_BACK) {
                        dialog.dismiss();
                        return true;
                    }
                    return false;
                });

                tvconfirm.setOnClickListener(v -> {
                    if (BWSApplication.isNetworkConnected(context)) {
                        showProgressBar();
                        Call<CardModel> listCall = APIClient.getClient().getRemoveCard(userId, listModel.getCustomer());
                        listCall.enqueue(new Callback<CardModel>() {
                            @Override
                            public void onResponse(Call<CardModel> call, Response<CardModel> response) {
                                hideProgressBar();
                                if (response.isSuccessful()) {
                                    CardModel cardModel = response.body();
                                    if (cardModel.getResponseCode().equalsIgnoreCase(getString(R.string.ResponseCodesuccess))) {
                                        prepareCardList();
                                    } else {
                                        BWSApplication.showToast(cardModel.getResponseMessage(), context);
                                    }
                                }
                            }

                            @Override
                            public void onFailure(Call<CardModel> call, Throwable t) {
                                hideProgressBar();
                            }
                        });
                    } else {
                        BWSApplication.showToast(getString(R.string.no_server_found), context);
                        hideProgressBar();
                    }
                });

                tvGoBack.setOnClickListener(v -> {
                    dialog.dismiss();
                });
                dialog.show();
                dialog.setCancelable(false);
            });
        }

        @Override
        public int getItemCount() {
            return listModelList.size();
        }

        public class MyViewHolder extends RecyclerView.ViewHolder {
            CardsListLayoutBinding binding;

            public MyViewHolder(CardsListLayoutBinding binding) {
                super(binding.getRoot());
                this.binding = binding;
            }
        }
    }


    private void hideProgressBar() {
        try {
            binding.progressBarHolder.setVisibility(View.GONE);
            binding.ImgV.setVisibility(View.GONE);
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void showProgressBar() {
        try {
            binding.progressBarHolder.setVisibility(View.VISIBLE);
            getWindow().setFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE, WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
            binding.ImgV.setVisibility(View.VISIBLE);
            binding.ImgV.invalidate();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onBackPressed() {
        Intent i = new Intent(PaymentActivity.this, OrderSummaryActivity.class);
        i.putExtra("comeFrom", "membership");
        i.putParcelableArrayListExtra("PlanData", listModelList2);
        i.putExtra("TrialPeriod", "");
        i.putExtra("position", position);
        startActivity(i);
        finish();
    }
}

File: app/src/main/java/com/brainwellnessspa/billingOrderModule/activities/PaymentActivity.kt
package com.brainwellnessspa.billingOrderModule.activities

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Application.ActivityLifecycleCallbacks
import android.app.Dialog
import android.app.NotificationManager
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.graphics.PorterDuff
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.*
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.brainwellnessspa.BWSApplication
import com.brainwellnessspa.BWSApplication.notificationId
import com.brainwellnessspa.billingOrderModule.models.CardListModel
import com.brainwellnessspa.billingOrderModule.models.CardModel
import com.brainwellnessspa.billingOrderModule.models.PayNowDetailsModel
import com.brainwellnessspa.billingOrderModule.models.PlanListBillingModel
import com.brainwellnessspa.billingOrderModule.models.SegmentPayment
import com.brainwellnessspa.R
import com.brainwellnessspa.addPaymentStripeModule.AddPaymentActivity
import com.brainwellnessspa.membershipModule.activities.StripeEnhanceMembershipUpdateActivity.Companion.renewPlanFlag
import com.brainwellnessspa.membershipModule.activities.StripeEnhanceMembershipUpdateActivity.Companion.renewPlanId
import com.brainwellnessspa.billingOrderModule.fragments.CurrentPlanFragment.Companion.PlanStatus
import com.brainwellnessspa.billingOrderModule.fragments.CurrentPlanFragment.Companion.invoicePayId
import com.brainwellnessspa.databinding.ActivityPaymentBinding
import com.brainwellnessspa.databinding.CardsListLayoutBinding
import com.brainwellnessspa.membershipModule.activities.StripeEnhanceOrderSummaryActivity
import com.brainwellnessspa.membershipModule.models.MembershipPlanListModel
import com.brainwellnessspa.services.GlobalInitExoPlayer.Companion.relesePlayer
import com.brainwellnessspa.utility.APINewClient
import com.brainwellnessspa.utility.CONSTANTS
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.segment.analytics.Properties
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class PaymentActivity : AppCompatActivity() {
    lateinit var binding: ActivityPaymentBinding
    var adapter: AllCardsAdapter? = null
    lateinit var context: Context
    var cardId: String? = ""
    var userId: String? = ""
    var trialPeriod: String? = ""
    var comeFrom: String? = ""
    var comesTrue: String? = ""
    val gson = Gson()
    var activity: Activity? = null
    var position = 0
    var listModelList2: ArrayList<PlanListBillingModel.ResponseData.Plan>? = null
    private var listModelList: ArrayList<MembershipPlanListModel.Plan>? = null
    private var numStarted = 0
    var stackStatus = 0
    var myBackPress = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_payment)
        context = this@PaymentActivity
        activity = this@PaymentActivity
        val shared = context.getSharedPreferences(CONSTANTS.PREFE_ACCESS_SIGNIN_COUSER, MODE_PRIVATE)
        userId = shared.getString(CONSTANTS.PREFE_ACCESS_UserId, "")
        val mLayoutManager: RecyclerView.LayoutManager = LinearLayoutManager(context)
        binding.rvCardList.layoutManager = mLayoutManager
        binding.rvCardList.itemAnimator = DefaultItemAnimator()
        if (intent != null) {
            if (intent.hasExtra("TrialPeriod")) {
                trialPeriod = intent.getStringExtra("TrialPeriod")
            }
            position = intent.getIntExtra("position", 0)
            if (intent.hasExtra("comeFrom")) {
                comeFrom = intent.getStringExtra("comeFrom")
                val json4 =  intent.getStringExtra("PlanData")
                val type1 = object : TypeToken<ArrayList<PlanListBillingModel.ResponseData.Plan>?>() {}.type
                listModelList2 = gson.fromJson(json4, type1)
            } else {
                val json4 =  intent.getStringExtra("PlanData")
                val type1 = object : TypeToken<ArrayList<MembershipPlanListModel.Plan>?>() {}.type
                listModelList = gson.fromJson(json4, type1)            }
        }
        if (intent != null) {
            comesTrue = intent.getStringExtra("ComesTrue")
        }
        binding.llBack.setOnClickListener {
            myBackPress = true
            val i = Intent(context, StripeEnhanceOrderSummaryActivity::class.java)
            if (intent.hasExtra("comeFrom")) {
                i.putExtra("PlanData", gson.toJson(listModelList2))
            }else{
                i.putExtra("PlanData", gson.toJson(listModelList))
            }
            i.putExtra("comeFrom", "membership")
            i.putExtra("ComesTrue", comesTrue)
            i.putExtra("TrialPeriod", "")
            i.putExtra("position", position)
            i.putExtra("Promocode", "")
            startActivity(i)
            finish()
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            registerActivityLifecycleCallbacks(AppLifecycleCallback())
        }
        binding.llAddNewCard.setOnClickListener {
            myBackPress = true
            if (BWSApplication.isNetworkConnected(context)) {
                val i = Intent(context, AddPaymentActivity::class.java)
                i.putExtra("ComePayment", "2")
                i.putExtra("ComesTrue", comesTrue)
                i.putExtra("comeFrom", "membership")
                i.putExtra("PlanData", gson.toJson(listModelList2))
                i.putExtra("TrialPeriod", "")
                i.putExtra("position", position)
                startActivity(i)
            } else {
                BWSApplication.showToast(getString(R.string.no_server_found), activity)
            }
        }
        //        prepareCardList();
    }

    public override fun onResume() {
        prepareCardList()
        super.onResume()
    }

    private fun prepareCardList() {
        try {
            if (BWSApplication.isNetworkConnected(context)) {
                BWSApplication.showProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                val listCall: Call<CardListModel> = APINewClient.client.getCardLists(userId)
                listCall.enqueue(object : Callback<CardListModel?> {
                    override fun onResponse(call: Call<CardListModel?>, response: Response<CardListModel?>) {
                        try {
                            val cardListModel: CardListModel? = response.body()
                            if (cardListModel!!.responseCode.equals(getString(R.string.ResponseCodesuccess))) {
                                BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                val p = Properties()
                                if (cardListModel.responseData!!.isEmpty()) {
                                    binding.rvCardList.adapter = null
                                    binding.rvCardList.visibility = View.GONE
                                    p.putValue("paymentCards", "")
                                } else {
                                    binding.rvCardList.visibility = View.VISIBLE
                                    adapter = AllCardsAdapter(cardListModel.responseData!!, binding.progressBar, binding.progressBarHolder, binding.rvCardList)
                                    binding.rvCardList.setAdapter(adapter)
                                    val section1: ArrayList<SegmentPayment> = ArrayList<SegmentPayment>()
                                    val e = SegmentPayment()
                                    val gson = Gson()
                                    for (i in cardListModel.responseData!!.indices) {
                                        e.cardId = (cardListModel.responseData!![i].customer)
                                        e.cardNumber = (getString(R.string.first_card_chars) + " " + cardListModel.responseData!![i].last4)
                                        e.cardHolderName = ("")
                                        e.cardExpiry = ("Valid: " + cardListModel.responseData!![i].expMonth.toString() + "/" + cardListModel.responseData!![i].expYear)
                                        section1.add(e)
                                    }
                                    p.putValue("paymentCards", gson.toJson(section1))
                                }
                                BWSApplication.addToSegment("Payment Screen Viewed", p, CONSTANTS.screen)
                                binding.btnCheckout.setOnClickListener {
                                    if (cardListModel.responseData!!.size == 0) {
                                        BWSApplication.showToast("Please enter card details", activity)
                                    } else {
                                        if (BWSApplication.isNetworkConnected(context)) {
                                            BWSApplication.showProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                            val listCall: Call<PayNowDetailsModel> = APINewClient.client.getPayNowDetails(userId, cardId, renewPlanId, renewPlanFlag, invoicePayId, PlanStatus)
                                            listCall.enqueue(object : Callback<PayNowDetailsModel?> {
                                                override fun onResponse(call: Call<PayNowDetailsModel?>, response: Response<PayNowDetailsModel?>) {
                                                    try {
                                                        if (response.isSuccessful) {
                                                            BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                                            val listModel1: PayNowDetailsModel? = response.body()
                                                            BWSApplication.showToast(listModel1!!.responseMessage, activity)
                                                            val i = Intent(context, BillingOrderActivity::class.java)
                                                            startActivity(i)
                                                            finish()
                                                        }
                                                    } catch (e: Exception) {
                                                        e.printStackTrace()
                                                    }
                                                }

                                                override fun onFailure(call: Call<PayNowDetailsModel?>, t: Throwable) {
                                                    BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                                }
                                            })
                                        } else {
                                            BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                            BWSApplication.showToast(getString(R.string.no_server_found), activity)
                                        }
                                    }
                                }
                            } else {
                                BWSApplication.showToast(cardListModel.responseMessage, activity)
                            }
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }

                    override fun onFailure(call: Call<CardListModel?>, t: Throwable) {
                        BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                    }
                })
            } else {
                BWSApplication.showToast(getString(R.string.no_server_found), activity)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    inner class AllCardsAdapter(var listModelList: List<CardListModel.ResponseData>,var ImgV: ProgressBar,var progressBarHolder: FrameLayout,var rvCardList: RecyclerView) : RecyclerView.Adapter<AllCardsAdapter.MyViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
            val v: CardsListLayoutBinding = DataBindingUtil.inflate(LayoutInflater.from(parent.context), R.layout.cards_list_layout, parent, false)
            return MyViewHolder(v)
        }

        @SuppressLint("ClickableViewAccessibility")
        override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
            val listModel: CardListModel.ResponseData = listModelList[position]
            val cardNo = getString(R.string.first_card_chars) + " " + listModel.last4
            holder.binding.tvCardNo.text = cardNo
            val expTime = "Valid: " + listModel.expMonth.toString() + "/" + listModel.expYear
            holder.binding.tvExpiryTime.text = expTime
            Glide.with(context).load(listModel.image).thumbnail(0.05f).diskCacheStrategy(DiskCacheStrategy.ALL).into(holder.binding.ivCardimg)
            if (listModel.isDefault.equals(CONSTANTS.FLAG_ONE)) {
                holder.binding.ivCheck.setImageResource(R.drawable.ic_checked_icon)
                cardId = listModel.customer
                val shared = getSharedPreferences(CONSTANTS.PREF_KEY_CardID, MODE_PRIVATE)
                val editor = shared.edit()
                editor.putString(CONSTANTS.PREF_KEY_CardID, cardId)
                editor.commit()
            } else {
                holder.binding.ivCheck.setImageResource(R.drawable.ic_unchecked_icon)
            }
            holder.binding.llAddNewCard.setOnClickListener {
                if (BWSApplication.isNetworkConnected(context)) {
                    BWSApplication.showProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                    val listCall: Call<CardListModel> = APINewClient.client.getChangeCard(userId, listModel.customer)
                    listCall.enqueue(object : Callback<CardListModel?> {
                        override fun onResponse(call: Call<CardListModel?>, response: Response<CardListModel?>) {
                            try {
                                BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                val cardListModel: CardListModel? = response.body()
                                if (cardListModel!!.responseCode.equals(getString(R.string.ResponseCodesuccess))) {
                                    if (cardListModel.responseData!!.isEmpty()) {
                                        rvCardList.adapter = null
                                        rvCardList.visibility = View.GONE
                                    } else {
                                        val p = Properties()
                                        p.putValue("cardId",listModel.customer)
                                        BWSApplication.addToSegment("Payment Card Selected", p, CONSTANTS.track)
                                        rvCardList.visibility = View.VISIBLE
                                        adapter = AllCardsAdapter(cardListModel.responseData!!, ImgV, progressBarHolder, rvCardList)
                                        rvCardList.adapter = adapter
                                    }
                                    BWSApplication.showToast(cardListModel.responseMessage, activity)
                                } else {
                                    BWSApplication.showToast(cardListModel.responseMessage, activity)
                                }
                            } catch (e: Exception) {
                                e.printStackTrace()
                            }
                        }

                        override fun onFailure(call: Call<CardListModel?>, t: Throwable) {
                            BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                        }
                    })
                } else {
                    BWSApplication.showToast(getString(R.string.no_server_found), activity)
                    BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                }
            }
            holder.binding.rlRemoveCard.setOnClickListener {
                val dialog = Dialog(context)
                dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)
                dialog.setContentView(R.layout.delete_payment_card)
                dialog.window!!.setBackgroundDrawable(ColorDrawable(ContextCompat.getColor(context,R.color.dark_blue_gray)))
                dialog.window!!.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
                val tvGoBack = dialog.findViewById<TextView>(R.id.tvGoBack)
                val Btn = dialog.findViewById<Button>(R.id.Btn)
                Btn.text = "DELETE"
                dialog.setOnKeyListener { v: DialogInterface?, keyCode: Int, event: KeyEvent? ->
                    if (keyCode == KeyEvent.KEYCODE_BACK) {
                        dialog.dismiss()
                        return@setOnKeyListener true
                    }
                    false
                }
                Btn.setOnTouchListener { view1: View, event: MotionEvent ->
                    when (event.action) {
                        MotionEvent.ACTION_DOWN -> {
                            val views = view1 as Button
                            views.background.setColorFilter(0x77000000, PorterDuff.Mode.SRC_ATOP)
                            view1.invalidate()
                        }
                        MotionEvent.ACTION_UP -> {
                            if (BWSApplication.isNetworkConnected(context)) {
                                BWSApplication.showProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                val listCall: Call<CardModel> = APINewClient.client.getRemoveCard(userId, listModel.customer)
                                listCall.enqueue(object : Callback<CardModel?> {
                                    override fun onResponse(call: Call<CardModel?>, response: Response<CardModel?>) {
                                        BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                        try {
                                            if (response.isSuccessful) {
                                                val cardModel: CardModel? = response.body()
                                                if (cardModel!!.responseCode.equals(getString(R.string.ResponseCodesuccess))) {
                                                    val p = Properties()
                                                    p.putValue("cardId",listModel.customer)
                                                    BWSApplication.addToSegment("Payment Card Removed", p, CONSTANTS.track)
                                                    prepareCardList()
                                                    dialog.dismiss()
                                                    BWSApplication.showToast(cardModel.responseMessage, activity)
                                                } else {
                                                    BWSApplication.showToast(cardModel.responseMessage, activity)
                                                }
                                            }
                                        } catch (e: Exception) {
                                            e.printStackTrace()
                                        }
                                    }

                                    override fun onFailure(call: Call<CardModel?>, t: Throwable) {
                                        BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                                    }
                                })
                            } else {
                                BWSApplication.showToast(getString(R.string.no_server_found), activity)
                                BWSApplication.hideProgressBar(binding.progressBar, binding.progressBarHolder, activity)
                            }
                            run {
                                val views = view1 as Button
                                views.background.clearColorFilter()
                                views.invalidate()
                            }
                        }
                        MotionEvent.ACTION_CANCEL -> {
                            val views = view1 as Button
                            views.background.clearColorFilter()
                            views.invalidate()
                        }
                    }
                    true
                }
                tvGoBack.setOnClickListener { v: View? -> dialog.dismiss() }
                dialog.show()
                dialog.setCancelable(false)
            }
        }

        override fun getItemCount(): Int {
            return listModelList.size
        }

        inner class MyViewHolder(var binding: CardsListLayoutBinding) : RecyclerView.ViewHolder(binding.root)
    }

    override fun onBackPressed() {
        myBackPress = true
        val i = Intent(context, StripeEnhanceOrderSummaryActivity::class.java)
        i.putExtra("comeFrom", "membership")
        i.putExtra("ComesTrue", comesTrue)
        if (intent.hasExtra("comeFrom")) {
            i.putExtra("PlanData", gson.toJson(listModelList2))
        }else{
            i.putExtra("PlanData", gson.toJson(listModelList))
        }
        i.putExtra("TrialPeriod", "")
        i.putExtra("position", position)
        i.putExtra("Promocode", "")
        startActivity(i)
        finish()
    }

    internal inner class AppLifecycleCallback : ActivityLifecycleCallbacks {
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
        override fun onActivityStarted(activity: Activity) {
            if (numStarted == 0) {
                stackStatus = 1
                Log.e("APPLICATION", "APP IN FOREGROUND")
                //app went to foreground
            }
            numStarted++
        }

        override fun onActivityResumed(activity: Activity) {}
        override fun onActivityPaused(activity: Activity) {}
        override fun onActivityStopped(activity: Activity) {
            numStarted--
            if (numStarted == 0) {
                if (!myBackPress) {
                    Log.e("APPLICATION", "Back press false")
                    stackStatus = 2
                } else {
                    myBackPress = true
                    stackStatus = 1
                    Log.e("APPLICATION", "back press true ")
                }
                Log.e("APPLICATION", "App is in BACKGROUND")
                // app went to background
            }
        }

        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
        override fun onActivityDestroyed(activity: Activity) {
            if (numStarted == 0 && stackStatus == 2) {
                Log.e("Destroy", "Activity Destoryed")
                val notificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
                notificationManager.cancel(notificationId)
                relesePlayer(applicationContext)
            } else {
                Log.e("Destroy", "Activity go in main activity")
            }
        }
    }
}